ext/opentelemetry-ext-asgi/src/opentelemetry/ext/asgi/__init__.py

```diff
@@ -0,0 +1,206 @@
+        span.set_status(Status(http_status_to_canonical_code(status_code)))
+
+
+def get_default_span_name(scope):
+    """Calculates a (generic) span name for an incoming HTTP request based on the ASGI scope."""
+
+    # TODO: Update once
+    #  https://github.com/open-telemetry/opentelemetry-specification/issues/270
+    #  is resolved
+    return scope.get("path", "/")
+
+
+class OpenTelemetryMiddleware:
+    """The ASGI application middleware.
+
+    This class is an ASGI middleware that starts and annotates spans for any
+    requests it is invoked with.
+
+    Args:
+        app: The ASGI application callable to forward requests to.
+    """
+
+    def __init__(self, app):
+        self.app = guarantee_single_callable(app)
+        self.tracer = trace.tracer_source().get_tracer(__name__, __version__)
+
+    async def __call__(self, scope, receive, send):
+        """The ASGI application
+
+        Args:
+            scope: A ASGI environment.
+            receive: An awaitable callable yielding dictionaries
+            send: An awaitable callable taking a single dictionary as argument.
+        """
+
+        parent_span = propagators.extract(get_header_from_scope, scope)
+        span_name = get_default_span_name(scope)
+
+        with self.tracer.start_as_current_span(
+            span_name,
+            parent_span,
+            kind=trace.SpanKind.SERVER,
+            attributes=collect_request_attributes(scope),
+        ):
+
+            @wraps(receive)
+            async def wrapped_receive():
+                with self.tracer.start_as_current_span(
+                    span_name + " (unknown-receive)"
+                ) as receive_span:
+                    payload = await receive()
+                    if payload["type"] == "websocket.receive":
+                        set_status_code(receive_span, 200)
+                        receive_span.set_attribute(
+                            "http.status_text", payload["text"]
+                        )
+
+                    receive_span.update_name(
+                        span_name + " (" + payload["type"] + ")"
+                    )
+                    receive_span.set_attribute("type", payload["type"])
+                return payload
+
+            @wraps(send)
+            async def wrapped_send(payload):
+                with self.tracer.start_as_current_span(
+                    span_name + " (unknown-send)"
+                ) as send_span:
+                    if payload["type"] == "http.response.start":
+                        status_code = payload["status"]
+                        set_status_code(send_span, status_code)
+                    elif payload["type"] == "websocket.send":
+                        set_status_code(send_span, 200)
+                        send_span.set_attribute(
+                            "http.status_text", payload["text"]
+                        )
+
+                    send_span.update_name(
```

Comment:
update_name is highly discouraged, and it's on my personal list to try to petition to remove it: https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/api-tracing.md#updatename

It looks like it's possible to refactor this to figure out the span name before we create the span.

Reply:
It is indeed possible to set the right span name from the get-go, but only for the send function. For the receive function we do not know what we are receiving, until we actually process the payload.

The rationale to use update_name in send, is simply to keep the two functions similar in structure. I will change it if you'd like me to, but I'm not quite sure how to adjust the receiver to avoid update_name.
- My only suggestion would be to have very generic send and receive span names, and then eventually add attributes about the nature of the payload as it comes in.

Reply:
I'd vote for just using the right span name in send from the beginning. It's not clear to me what type of behavior will be affected when update_name is used, but as I understand it there will be pre-aggregations and optimizations that will group things by the original name. As such, minimizing update_name should probably be a higher priority than a little bit of code cohesiveness.

Reply:
I will move the naming scheme up for send, and keep it as is for receive.

Reply:
Resolved in https://github.com/open-telemetry/opentelemetry-python/pull/402/commits/4a960608314abe107b40106acbddc3163a76a2d8

Reply:
Changed in: https://github.com/open-telemetry/opentelemetry-python/pull/402/commits/9c4242d742bb91b5727d9c2c7408a324c999b86c